Stage 1

Icuisset opened the pull request:
Merge Stage 1 into main in order to publish on server

AygulMalikova commented:
Unfortunately, your work is rejected without review because when I try to run `npm run dev` or `npm run start` there is an error in the console (MongooseServerSelectionError: Could not connect to any servers in your MongoDB Atlas cluster), and the application crashes. Please name sure that when running the application locally your app is connected to the local mongodb instead of the remote one. Also, please note that you need to send to review not the link to the branch but the link to the pul-request. And finally, make sure that your .env file is not presented in your repository.

Icuisset commented:
Hello,

Apologies , I changed properties to my MongoDb atlas cluster ( as in Sprint 15 , I'm using this because this is the method I want to use in my future live app ) and I removed .env.

I hope everything will work properly now.
Here is the link to the pull-request url :
https://github.com/Icuisset/news-explorer-api/pull/1

Kind regards,

Isabelle.

From: AygulMalikova ***@***.***>
Sent: 26 July 2021 19:28
To: Icuisset/news-explorer-api ***@***.***>
Cc: Icuisset ***@***.***>; Author ***@***.***>
Subject: Re: [Icuisset/news-explorer-api] Stage 1 (#1)


Unfortunately, your work is rejected without review because when I try to run npm run dev or npm run start there is an error in the console (MongooseServerSelectionError: Could not connect to any servers in your MongoDB Atlas cluster), and the application crashes. Please name sure that when running the application locally your app is connected to the local mongodb instead of the remote one. Also, please note that you need to send to review not the link to the branch but the link to the pul-request. And finally, make sure that your .env file is not presented in your repository.

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIcuisset%2Fnews-explorer-api%2Fpull%2F1%23issuecomment-886929583&data=04%7C01%7C%7C2cb0322f51154ca749fe08d950631ad6%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637629208853562636%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=ogzkgEy6gHzXjPPKcOAFcZYwFMf4AEkq8IE%2F8B5O0rA%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FARFCR3DYRGFLSO3HDNWMWTLTZWSLFANCNFSM5BAF6TFQ&data=04%7C01%7C%7C2cb0322f51154ca749fe08d950631ad6%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637629208853562636%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=dyA4SjOOBm9FL7nV2iPjMnpJCvNTwBl0809Z1sYYGDo%3D&reserved=0>.

AygulMalikova commented:
# Summary:
## Backend

### Performance Criteria
- [x] <!--3.64--> The repository contains all the necessary infrastructure files:
    - [x] A  package.json file.
    - [x] An .editorconfig file.
    - [x] An .eslintrc file, which extends the airbnb-base configuration, and devDependencies required for the linter.
    - [x] A .gitignore file.
    - [x] Additionally, an exception for `_id` is added in .eslintrc. The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
- [x] <!--3.64--> No linting errors.
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
    - [x] An `npm run start` command that starts the server on `localhost:3000`.
    - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [x] <!--12.57--> When all dependencies are installed, the application starts with npm run dev without errors.
- [ ] <!--25.45--> The following routes function as described:
    - [x] A request to `GET /users/me` returns information about the user (email and name).
    - [ ] `GET /articles` returns all articles saved by the user.
      **When I make this request in the response there are also articles that are created by other users (https://prnt.sc/1gj8zbc). Also, make sure that when a new user request his articles (before creating new ones) there should an empty array**
    - [x] `POST /articles` creates an article with the data passed inside the request body.
    - [ ] `DELETE /articles/articleId` deletes the saved article using `_id`.
      **When I try to remove an article of another user there is the error 'Article ID is not valid' but anyway card is removed**
    - [x] `POST /signup` creates a user with the data passed inside the request body.
    - [x] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--3.64--> API errors are handled:
    - [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
    - [x] Asynchronous handlers end with a `catch()` block.
    - [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> In production mode, the database address is taken from `process.env`.
  **Please make sure that when running the application locally your app is connected to the local mongodb ('mongodb://localhost:27017/newsdb') instead of the remote one.**
- [x] <!--3.64--> Safe password storage has been implemented:
    - [x] Passwords are stored in an encrypted format.
    - [x] The API does not return a password hash to the client.
- [x] <!--3.64--> Data is validated before being added to the database.
- [x] <!--3.64--> Users can't delete saved cards from other user profiles.
- [x] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md. 
- [x] <!--3.64--> Storing the private key for creating a JWT is implemented correctly:
    - [x] For the production build, it is stored in an `.env` file, and this file should not be added to Git.
    - [x] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine and an error won't occur if there is no `.env` file present.

### Best Practices
- [x] <!--1.5--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--1.5--> Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
- [x] <!--1.5--> Logging is set up:
    - [x] All requests and responses are logged to the `request.log` file.
    - [x] All errors are logged to the `error.log` file.
    - [x] Log files aren't added to the git repository.
- [x] <!--1.5--> Errors are handled by a centralized handler.
- [x] <!--1.5--> Centralized error handling is described inside a separate module.
- [x] <!--1.5--> In case of an error, the API returns a response status that matches the error type.
- [x] <!--1.5--> The server can be accessed via HTTPS using the domain specified in `README.md`. 
- [ ] <!--1.5--> The application API is located on a domain with a name of the following format: name.zone/api (not just name.zone).
    - [ ] Correct: news-explorer.tk/api
    - [ ] Incorrect: news-explorer.tk
- [ ] <!--1.5--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.

### Recommendations
- [x] <!--0.83--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--0.83--> The Helmet module is used to set security-related headers.
- [ ] <!--0.83--> Configuration and constants are stored in separate files:
    - [ ] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
    - [ ] Application constants (response and error messages) are stored inside a separate file with constants.
- [ ] <!--0.83--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [ ] <!--0.83--> The rate limiter is configured in a separate file and imported into app.js.
- [x] <!--0.83--> The API is hosted on a separate subdomain, e.g. api.news-explorer.tk.

## Number of points:
69

## Additional comments:
Great job! 

You have a well-structured and clean code, all the necessary functionality is implemented!
But there are some small remarks. First of all, I ask you to check that `GET /articles` and `DELETE /articles/articleId` work as required. And also, other comments were left to improve your code (see the checklist).

I remind you that the work will be accepted as soon as you receive at least 80 points. Wish you luck!

Icuisset commented:
Hello,
Thank you so much for all your super helpful comments, I just resubmitted ( after uploading the production version also to https://api.news-explorer.info )
Kind regards,
Isabelle
________________________________
From: AygulMalikova ***@***.***>
Sent: Tuesday, July 27, 2021 9:38:12 AM
To: Icuisset/news-explorer-api ***@***.***>
Cc: Icuisset ***@***.***>; Author ***@***.***>
Subject: Re: [Icuisset/news-explorer-api] Stage 1 (#1)


Summary:
Backend
Performance Criteria

  *    The repository contains all the necessary infrastructure files:
     *    A package.json file.
     *    An .editorconfig file.
     *    An .eslintrc file, which extends the airbnb-base configuration, and devDependencies required for the linter.
     *    A .gitignore file.
     *    Additionally, an exception for _id is added in .eslintrc. The following rules are forbidden: eslint-disable, eslint-disable-line, and eslint-disable-next-line
  *    No linting errors.
  *    The scripts section of the package.json file contains the following:
     *    An npm run start command that starts the server on localhost:3000.
     *    An npm run dev command that starts the server on localhost:3000 with hot reloading.
  *    When all dependencies are installed, the application starts with npm run dev without errors.
  *    The following routes function as described:
     *    A request to GET /users/me returns information about the user (email and name).
     *    GET /articles returns all articles saved by the user.
When I make this request in the response there are also articles that are created by other users (https://prnt.sc/1gj8zbc<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fprnt.sc%2F1gj8zbc&data=04%7C01%7C%7C46e0ae2044fe433950a008d950d9df07%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637629718953377632%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=euroZdudu0%2FCBBxXByInTAgsV0iGjC%2FhIKkOBevuTFc%3D&reserved=0>). Also, make sure that when a new user request his articles (before creating new ones) there should an empty array
     *    POST /articles creates an article with the data passed inside the request body.
     *    DELETE /articles/articleId deletes the saved article using _id.
When I try to remove an article of another user there is the error 'Article ID is not valid' but anyway card is removed
     *    POST /signup creates a user with the data passed inside the request body.
     *    POST /signin returns a JWT when the correct email and password are passed in the request body.
  *    All routes are protected with authorization, except for /signin and /signup.
  *    User routes and article routes are described in separate files.
  *    API errors are handled:
     *    If something is wrong with the request, the server returns a response with an error message and a corresponding status.
     *    Asynchronous handlers end with a catch() block.
     *    The API does not return standard database or Node.js errors.
  *    In production mode, the database address is taken from process.env.
Please make sure that when running the application locally your app is connected to the local mongodb ('mongodb://localhost:27017/newsdb') instead of the remote one.
  *    Safe password storage has been implemented:
     *    Passwords are stored in an encrypted format.
     *    The API does not return a password hash to the client.
  *    Data is validated before being added to the database.
  *    Users can't delete saved cards from other user profiles.
  *    The server can be accessed via HTTPS using the domain specified in README.md.
  *    Storing the private key for creating a JWT is implemented correctly:
     *    For the production build, it is stored in an .env file, and this file should not be added to Git.
     *    In development mode (process.env.NODE_ENV !== 'production'), the code runs and works fine and an error won't occur if there is no .env file present.

Best Practices

  *    Asynchronous operations are implemented using promises or async/await.
  *    Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
  *    Logging is set up:
     *    All requests and responses are logged to the request.log file.
     *    All errors are logged to the error.log file.
     *    Log files aren't added to the git repository.
  *    Errors are handled by a centralized handler.
  *    Centralized error handling is described inside a separate module.
  *    In case of an error, the API returns a response status that matches the error type.
  *    The server can be accessed via HTTPS using the domain specified in README.md.
  *    The application API is located on a domain with a name of the following format: name.zone/api (not just name.zone).
     *    Correct: news-explorer.tk/api
     *    Incorrect: news-explorer.tk
  *    All routes are connected to the index.js file, which is located in the routes folder, and app.js contains one main route handled by routes.

Recommendations

  *    For API errors, classes have been created to extend the Error constructor.
  *    The Helmet module is used to set security-related headers.
  *    Configuration and constants are stored in separate files:
     *    The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
     *    Application constants (response and error messages) are stored inside a separate file with constants.
  *    A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
  *    The rate limiter is configured in a separate file and imported into app.js.
  *    The API is hosted on a separate subdomain, e.g. api.news-explorer.tk.

Number of points:

69

Additional comments:

Great job!

You have a well-structured and clean code, all the necessary functionality is implemented!
But there are some small remarks. First of all, I ask you to check that GET /articles and DELETE /articles/articleId work as required. And also, other comments were left to improve your code (see the checklist).

I remind you that the work will be accepted as soon as you receive at least 80 points. Wish you luck!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIcuisset%2Fnews-explorer-api%2Fpull%2F1%23issuecomment-887324621&data=04%7C01%7C%7C46e0ae2044fe433950a008d950d9df07%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637629718953377632%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=T15QdUr1fCNm0d6mp6u9BiCkfgX2Dm6zOXqM%2BL4oiEs%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FARFCR3CZIPG2L2H3F5PUFRLTZZV7JANCNFSM5BAF6TFQ&data=04%7C01%7C%7C46e0ae2044fe433950a008d950d9df07%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637629718953387590%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=BPgLZuIxsy%2Bupwddiq8cN%2FSgZ3rYnNqNvIkwstlLx5Y%3D&reserved=0>.

AygulMalikova commented:
# Summary:
## Backend

### Performance Criteria
- [x] <!--3.64--> The repository contains all the necessary infrastructure files:
  - [x] A  package.json file.
  - [x] An .editorconfig file.
  - [x] An .eslintrc file, which extends the airbnb-base configuration, and devDependencies required for the linter.
  - [x] A .gitignore file.
  - [x] Additionally, an exception for `_id` is added in .eslintrc. The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
- [x] <!--3.64--> No linting errors.
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
  - [x] An `npm run start` command that starts the server on `localhost:3000`.
  - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [x] <!--12.57--> When all dependencies are installed, the application starts with npm run dev without errors.
- [x] <!--25.45--> The following routes function as described:
	- [x] A request to `GET /users/me` returns information about the user (email and name).
	- [x] `GET /articles` returns all articles saved by the user.
	- [x] `POST /articles` creates an article with the data passed inside the request body.
	- [x] `DELETE /articles/articleId` deletes the saved article using `_id`.
	- [x] `POST /signup` creates a user with the data passed inside the request body.
	- [x] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--3.64--> API errors are handled:
	- [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
	- [x] Asynchronous handlers end with a `catch()` block.
	- [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> In production mode, the database address is taken from `process.env`.
- [x] <!--3.64--> Safe password storage has been implemented:
	- [x] Passwords are stored in an encrypted format.
	- [x] The API does not return a password hash to the client.
- [x] <!--3.64--> Data is validated before being added to the database.
- [x] <!--3.64--> Users can't delete saved cards from other user profiles.
- [x] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md. 
- [x] <!--3.64--> Storing the private key for creating a JWT is implemented correctly:
	- [x] For the production build, it is stored in an `.env` file, and this file should not be added to Git.
	- [x] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine and an error won't occur if there is no `.env` file present.

### Best Practices
- [x] <!--1.5--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--1.5--> Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
- [x] <!--1.5--> Logging is set up:
	- [x] All requests and responses are logged to the `request.log` file.
	- [x] All errors are logged to the `error.log` file.
	- [x] Log files aren't added to the git repository.
- [x] <!--1.5--> Errors are handled by a centralized handler.
- [x] <!--1.5--> Centralized error handling is described inside a separate module.
- [x] <!--1.5--> In case of an error, the API returns a response status that matches the error type.
- [x] <!--1.5--> The server can be accessed via HTTPS using the domain specified in `README.md`. 
- [ ] <!--1.5--> The application API is located on a domain with a name of the following format: name.zone/api (not just name.zone).
	- [ ] Correct: news-explorer.tk/api
	- [ ] Incorrect: news-explorer.tk
- [ ] <!--1.5--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.

### Recommendations
- [x] <!--0.83--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--0.83--> The Helmet module is used to set security-related headers.
- [ ] <!--0.83--> Configuration and constants are stored in separate files:
	- [ ] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
	- [ ] Application constants (response and error messages) are stored inside a separate file with constants.
- [ ] <!--0.83--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [ ] <!--0.83--> The rate limiter is configured in a separate file and imported into app.js.
- [x] <!--0.83--> The API is hosted on a separate subdomain, e.g. api.news-explorer.tk.

## Number of points:
95

## Additional comments:
Congratulations, you have successfully completed the first stage of your final project! In your project, there is still an opportunity to improve it, and it would be great if you fix them yourself (see the checklist). I was glad to check your work. I wish you success!
